snapcraft/storeapi/__init__.py

```diff
@@ -13,9 +13,282 @@
+        super().__init__(name=name, channel=channel, arch=arch)
+
+
+class SHAMismatchError(StoreError):
+
+    fmt = 'SHA512 checksum for {path} is not {expected_sha}.'
+
+    def __init__(self, path, expected_sha):
+        super().__init__(path=path, expected_sha=expected_sha)
+
+
+def _macaroon_auth(conf, acl):
+    """Format a macaroon and its associated discharge.
+
+    :return: A string suitable to use in an Authorization header.
+    """
+    macaroon, discharge = conf.get_macaroon(acl)
+    auth = 'Macaroon root={}, discharge={}'.format(macaroon, discharge)
+    return auth
+
+
+class SCAClient(object):
+    """High-level client for the V2.0 API SCA resources."""
+
+    def __init__(self):
+        self.conf = config.Config()
+        self.session = requests.Session()
+        self.root_url = os.environ.get('UBUNTU_STORE_API_ROOT_URL',
+                                       UBUNTU_STORE_API_ROOT_URL)
+        sso_url = os.environ.get(
+            'UBUNTU_SSO_API_ROOT_URL', UBUNTU_SSO_API_ROOT_URL)
+        self.sso = sso.V2ApiClient(sso_url)
+        # Will be set by upload()
+        self.updown = None
+        # Will be set by download()
+        self.cpi = None
+
+    # API for snapcraft
+
+    def login(self, email, password, one_time_password=None):
+        result = dict(success=False, body=None)
+        acls = ('store_read', 'store_write',
+                'package_access', 'package_upload')
+        macaroons = {}
+        for acl in acls:
+            macaroon, error = self.get_macaroon(acl)
+            if error:
+                result['errors'] = error
+                break
+            macaroons[acl] = macaroon
+        else:
+            discharges, error = self.get_discharges(
+                email, password, one_time_password, macaroons)
+            if error:
+                result['errors'] = error
+            else:
+                # All macaroons have been discharged, save them in the config
+                for acl in acls:
+                    self.conf.set(acl, ','.join([macaroons[acl],
+                                                 discharges[acl]]))
+                self.conf.save()
+                result['success'] = True
+        return result
+
+    def logout(self):
+        self.conf.clear()
+        self.conf.save()
+
+    def register_name(self, name):
+        data = dict(snap_name=name, series=DEFAULT_SERIES)
+        auth = _macaroon_auth(self.conf, 'package_upload')
+        response = self.post('register-name/',
+                             data=json.dumps(data),
+                             headers={'Authorization': auth,
+                                      'Content-Type': 'application/json'})
+        if not response.ok:
+            # if (response['errors'] == ['Authorization Required']
+            #     and (response.headers['WWW-Authenticate']
+            #          == "Macaroon needs_refresh=1":
+            # Refresh the discharge macaroon and retry
```

Comment:
This is tricky ;-)
As mentioned outside this MP, there are other pieces to complete the macaroons implementation (mostly publish and here, macaroon refresh).
I did my best to limit this MP to the minimal set of features allowing snapcraft to switch from OAuth to macaroons.
Sooo, it's hard to file bugs until this MP has landed since they would refer to stuff that is not yet part of the project.
On that topic, I have the same dilemma about two issues I fixed for the integration coverage since they are against code that is not part of the project anymore...